From: Stefan Eissing <stefan@eissing.org>
Date: Fri, 21 Aug 2026 11:50:33 +0200
Subject: ws: fix write callback error handling

When websocket writes out data thought the registered write callback and
that returns an error, clear the write buffers, so a sub-sequent flush
will not attempt another write. The transfer is failed at that point and
should no longer invoke the callback.

Fixes #22627
Reported-by: Scott Talbert
Closes #22634

Origin: upstream, https://github.com/curl/curl/commit/7e7ee16dd3a6e36029fd41c595724c6614176bcb
Last-Update: 2026-08-24
---
 lib/ws.c                                           |  6 +-
 tests/http/test_20_websockets.py                   | 15 ++++-
 tests/libtest/Makefile.inc                         |  1 +
 .../libtest/{cli_ws_pause.c => cli_ws_write_err.c} | 75 ++++++----------------
 4 files changed, 39 insertions(+), 58 deletions(-)
 copy tests/libtest/{cli_ws_pause.c => cli_ws_write_err.c} (58%)

diff --git a/lib/ws.c b/lib/ws.c
index 25ac9dc..4d6b988 100644
--- a/lib/ws.c
+++ b/lib/ws.c
@@ -754,7 +754,8 @@ static CURLcode ws_cw_write(struct Curl_easy *data,
     }
     else if(result) {
       failf(data, "[WS] decode payload error %d", (int)result);
-      return result;
+      Curl_bufq_reset(&ctx->buf);
+      goto out;
     }
   }
 
@@ -812,7 +813,8 @@ static CURLcode ws_cw_flush(struct Curl_easy *data,
       }
       else if(result) {
         failf(data, "[WS] decode payload error %d", (int)result);
-        return result;
+        Curl_bufq_reset(&ctx->buf);
+        goto out;
       }
     }
   }
diff --git a/tests/http/test_20_websockets.py b/tests/http/test_20_websockets.py
index 8f8b2ef..c859279 100644
--- a/tests/http/test_20_websockets.py
+++ b/tests/http/test_20_websockets.py
@@ -296,7 +296,10 @@ class TestWebsockets:
             except OSError as e:
                 st["err"] = e
 
-        curl = CurlClient(env=env)
+        run_env = os.environ.copy()
+        if 'CURL_DEBUG' in run_env:
+            del run_env['CURL_DEBUG']
+        curl = CurlClient(env=env, run_env=run_env)
         send_rounds = 2
         threading.Thread(target=srv, daemon=True).start()
         while "p" not in st and "err" not in st:
@@ -344,3 +347,13 @@ class TestWebsockets:
         url = f'ws://localhost:{ws_4frames.port}/large'
         r = client.run(args=[url, payload])
         r.check_exit_code(0)
+
+    # test handling of write callback errors
+    def test_20_14_write_err(self, env: Env, ws_4frames):
+        payload = 127 * "x"
+        client = LocalClient(env=env, name='cli_ws_write_err')
+        if not client.exists():
+            pytest.skip(f'example client not built: {client.name}')
+        url = f'ws://localhost:{ws_4frames.port}/small'
+        r = client.run(args=[url, payload])
+        r.check_exit_code(0)
diff --git a/tests/libtest/Makefile.inc b/tests/libtest/Makefile.inc
index 6311fe0..1554f1b 100644
--- a/tests/libtest/Makefile.inc
+++ b/tests/libtest/Makefile.inc
@@ -66,6 +66,7 @@ TESTS_C = \
   cli_ws_data.c \
   cli_ws_pause.c \
   cli_ws_pingpong.c \
+  cli_ws_write_err.c \
   \
   lib500.c lib501.c lib502.c lib503.c lib504.c lib505.c lib506.c lib507.c \
   lib508.c lib509.c lib510.c lib511.c lib512.c lib513.c lib514.c lib515.c \
diff --git a/tests/libtest/cli_ws_pause.c b/tests/libtest/cli_ws_write_err.c
similarity index 58%
copy from tests/libtest/cli_ws_pause.c
copy to tests/libtest/cli_ws_write_err.c
index b423314..7018224 100644
--- a/tests/libtest/cli_ws_pause.c
+++ b/tests/libtest/cli_ws_write_err.c
@@ -27,22 +27,21 @@
 
 #ifndef CURL_DISABLE_WEBSOCKETS
 
-struct test_ws_pause_ctx {
+struct test_ws_write_err_ctx {
   CURL *easy;
   int callback_count;
-  int paused;
-  int frames;
   int errors;
   int closed;
 };
 
-static size_t test_ws_pause_write_cb(char *ptr, size_t size, size_t nmemb,
-                                     void *userdata)
+static size_t test_ws_write_err_write_cb(char *ptr, size_t size, size_t nmemb,
+                                         void *userdata)
 {
-  struct test_ws_pause_ctx *ctx = userdata;
+  struct test_ws_write_err_ctx *ctx = userdata;
   size_t nbytes = size * nmemb;
   const struct curl_ws_frame *meta = curl_ws_meta(ctx->easy);
 
+  (void)ptr;
   ctx->callback_count++;
   if(!meta) {
     curl_mfprintf(stderr, "write_cb: ERROR call #%d with meta=NULL\n",
@@ -50,54 +49,25 @@ static size_t test_ws_pause_write_cb(char *ptr, size_t size, size_t nmemb,
     ++ctx->errors;
     return CURL_WRITEFUNC_ERROR;
   }
-
-  ++ctx->frames;
-  if(meta->len < nbytes) {
-    curl_mfprintf(stderr, "write_cb: ERROR call #%d more data than current "
-                  "frame, FRAME[flags=0x%x age=%d offset=%" FMT_OFF_T
-                  " bytesleft=%" FMT_OFF_T " len=%zu], bytes=%zu\n",
-                  ctx->callback_count,
-                  (unsigned int)meta->flags, meta->age,
-                  meta->offset, meta->bytesleft, meta->len, nbytes);
-    ++ctx->errors;
-    return CURL_WRITEFUNC_ERROR;
-  }
-
-  if(meta->flags == 0x1) { /* TEXT frame */
-    curl_mfprintf(stderr, "write_cb: call #%d FRAME[TEXT age=%d offset=%"
-                  FMT_OFF_T " bytesleft=%" FMT_OFF_T " len=%zu] '%.*s'\n",
-                  ctx->callback_count, meta->age,
-                  meta->offset, meta->bytesleft, meta->len,
-                  (int)nbytes, ptr);
-  }
-  else if(meta->flags == 0x8) {
-    curl_mfprintf(stderr, "write_cb: call #%d FRAME[CLOSE age=%d offset=%"
-                  FMT_OFF_T " bytesleft=%" FMT_OFF_T " len=%zu] bytes=%zu\n",
-                  ctx->callback_count, meta->age,
-                  meta->offset, meta->bytesleft, meta->len, nbytes);
-    ctx->closed = TRUE;
-  }
-  else {
-    curl_mfprintf(stderr, "write_cb: call #%d FRAME[flags=0x%x age=%d offset=%"
-                  FMT_OFF_T " bytesleft=%" FMT_OFF_T " len=%zu\n",
-                  ctx->callback_count,
-                  (unsigned int)meta->flags, meta->age,
-                  meta->offset, meta->bytesleft, meta->len);
+  if(ctx->callback_count > 2) {
+    ctx->errors++;
+    curl_mfprintf(stderr, "write_cb: call #%d should not happen\n",
+                  ctx->callback_count);
+    return 0;
   }
-
-  if(ctx->callback_count == 1 || ctx->callback_count == 3) {
-    ctx->paused = 1;
-    curl_mfprintf(stderr, "write_cb: call #%d PAUSING\n", ctx->callback_count);
-    return CURL_WRITEFUNC_PAUSE;
+  if(ctx->callback_count > 1) {
+    curl_mfprintf(stderr, "write_cb: call #%d return error\n",
+                  ctx->callback_count);
+    return 0;
   }
   return nbytes;
 }
 #endif /* CURL_DISABLE_WEBSOCKETS */
 
-static CURLcode test_cli_ws_pause(const char *URL)
+static CURLcode test_cli_ws_write_err(const char *URL)
 {
 #ifndef CURL_DISABLE_WEBSOCKETS
-  struct test_ws_pause_ctx ctx;
+  struct test_ws_write_err_ctx ctx;
   CURLM *multi;
   int still_running = 0;
   int msgs_left = 0;
@@ -117,7 +87,8 @@ static CURLcode test_cli_ws_pause(const char *URL)
   }
 
   curl_easy_setopt(ctx.easy, CURLOPT_URL, URL);
-  curl_easy_setopt(ctx.easy, CURLOPT_WRITEFUNCTION, test_ws_pause_write_cb);
+  curl_easy_setopt(ctx.easy, CURLOPT_WRITEFUNCTION,
+                   test_ws_write_err_write_cb);
   curl_easy_setopt(ctx.easy, CURLOPT_WRITEDATA, &ctx);
   curl_easy_setopt(ctx.easy, CURLOPT_VERBOSE, 1L);
 
@@ -126,13 +97,6 @@ static CURLcode test_cli_ws_pause(const char *URL)
 
   while(still_running && !ctx.closed && !ctx.errors) {
 
-    if(ctx.paused) {
-      curl_mfprintf(stderr, "main: wait and UNPAUSE\n");
-      curlx_wait_ms(500);
-      ctx.paused = 0;
-      curl_easy_pause(ctx.easy, CURLPAUSE_CONT);
-    }
-
     curl_mfprintf(stderr, "main: poll\n");
     curl_multi_poll(multi, NULL, 0, 100, NULL);
     curl_mfprintf(stderr, "main: perform\n");
@@ -161,7 +125,8 @@ out:
     curl_multi_cleanup(multi);
   curl_global_cleanup();
 
-  return ctx.errors ? CURLE_WRITE_ERROR : CURLE_OK;
+  return (ctx.errors || (ctx.callback_count != 2)) ?
+         CURLE_WRITE_ERROR : CURLE_OK;
 
 #else /* !CURL_DISABLE_WEBSOCKETS */
   (void)URL;
